numba/cuda/cudadrv/driver.py

```diff
@@ -2715,6 +2735,88 @@ def complete(self):
         return cubin, size
 
 
+class NvrtcProgram:
+    """NvrtcProgram is for the managing the lifetime of nvrtcProgram objects.
+
+    If an error occurs during an NVRTC call, an exception is raised. When an
+    instance of this class is deleted, it attempts to delete the underlying
+    nvrtcProgram."""
+
+    def __init__(self, src, name):
+        # Create an nvrtcProgram
+        err, program = nvrtc.nvrtcCreateProgram(src, name.encode(), 0, [], [])
+        self.check(err)
+        self._program = program
+
+        with driver.get_active_context() as ac:
+            dev = driver.get_device(ac.devnum)
+            major, minor = dev.compute_capability
+
+        # Compilation options:
+        # - Compile for the current device's compute capability.
+        # - The CUDA include path is added.
+        # - Relocatable Device Code (rdc) is needed to prevent device functions
+        #   being optimized away.
+        arch = f'--gpu-architecture=compute_{major}{minor}'.encode()
+        include = f'-I{config.CUDA_INCLUDE_PATH}'.encode()
+        opts = [arch, include, b'-rdc', b'true']
+
+        # Compile the program
+        err, = nvrtc.nvrtcCompileProgram(self._program, len(opts), opts)
+
+        # First check whether the call failed due to a "normal" compiler error
+        compile_error = (err == nvrtc.nvrtcResult.NVRTC_ERROR_COMPILATION)
+        if not compile_error:
+            # Check for any other error
+            self.check(err)
+
+        # Get log from compilation
+        err, log_size = nvrtc.nvrtcGetProgramLogSize(self._program)
+        self.check(err)
+        log = b' ' * log_size
+        err, = nvrtc.nvrtcGetProgramLog(self._program, log)
+        self.check(err)
+
+        # If the compile failed, provide the log in an exception
+        if compile_error:
+            msg = (f'NVRTC Compilation failure compiling {name}:\n\n'
+                   f'{log.decode()}')
+            raise NvrtcError(msg)
+
+        # Otherwise, if there's any content in the log, present it as a warning
+        if log_size > 1:
+            msg = f"NVRTC log messages compiling {name}:\n\n{log.decode()}"
+            warnings.warn(msg)
```

Comment:
Should this be a Numba CUDA specific warning class?

Reply:
None of the other warnings in driver.py use a specific class. I think there's not too great a need for this - I'd imagine you always want to resolve the warnings rather than filter them, but I'm open to a suggestion as to what I might be missing / misjudging here.

Reply:
ok, thanks, I think for consistency with the current code base it's fine to leave this as is in this PR, but might be something to consider at a later time.